### PR TITLE
executor: fix memory corrupt in COUNT/JSON_OBJECTAGG/GROUP_CONCAT

### LINKED ISSUE #17148: Incorrect results with count(distinct string)
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

Use mysql test data
https://dev.mysql.com/doc/employee/en/employees-installation.html

```mysql
mysql> select count(distinct first_name) from employees where hire_date between '1986-01-01' and '1986-12-31';
+----------------------------+
| count(distinct first_name) |
+----------------------------+
|                      24761 |
+----------------------------+
1 row in set (0.02 sec)

mysql> select count(distinct first_name) from employees where hire_date between '1986-01-01' and '1986-12-31';
+----------------------------+
| count(distinct first_name) |
+----------------------------+
|                      24760 |
+----------------------------+
1 row in set (0.01 sec)

```

### 2. What did you expect to see?

```mysql
mysql> select count(distinct first_name) from employees where hire_date between '1986-01-01' and '1986-12-31';
+----------------------------+
| count(distinct first_name) |
+----------------------------+
|                       1275 |
+----------------------------+
1 row in set (0.00 sec)
```
### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)

v4.0.0-rc.1

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve?

Issue Number: close #17148 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

Add a unit test, corrupt the original `chunk` memory after `UpdatePartialResult` is called. It exposes two more problem

1. Fix memory reference problem of `count(distinct <string>)`, It's a bug introduced by https://github.com/pingcap/tidb/pull/15323
2. Fix memory reference problem of `JSON_OBJECTAGG`, exposed by the unit test.
3. Fix memory reference problem of `GROUP_CONCAT(ORDER BY <string>)`, exposed by the unit test.

Copy the string to avoid memory reference.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

Side effects

- None

### Release note 

- fix potential incorrect results in COUNT/JSON_OBJECTAGG/GROUP_CONCAT with varchar arguments.

### CODE REVIEW

**Comment by zz-jason** [comment]
Would it impact the performance of these functions?

File: executor/aggfuncs/func_count_distinct.go [hunk 1]
```diff
@@ -254,6 +255,7 @@ func (e *countOriginalWithDistinct4String) UpdatePartialResult(sctx sessionctx.C
 		if p.valSet.Exist(input) {
 			continue
 		}
+		input = stringutil.Copy(input)
```
Comment (zz-jason) [inline executor/aggfuncs/func_count_distinct.go:258]: how about modifying the `stringmap`?

Comment (zz-jason) [reply executor/aggfuncs/func_count_distinct.go:258]:
```sh
[jianzhang.zj:~/Code/tidb] git:(master ✔)
➜ grepgo "set.StringSet" | grep -v "test" | grep -v "(" | grep -v "{" | grep "set.StringSet"
./planner/core/planbuilder.go:354:      underlyingViewNames set.StringSet
./planner/core/memtable_predicate_extractor.go:183:     result set.StringSet,
./planner/core/memtable_predicate_extractor.go:472:     excludeCols set.StringSet,
./planner/core/memtable_predicate_extractor.go:482:             var values set.StringSet
./planner/core/memtable_predicate_extractor.go:513:     NodeTypes set.StringSet
./planner/core/memtable_predicate_extractor.go:519:     Instances set.StringSet
./planner/core/memtable_predicate_extractor.go:566:     NodeTypes set.StringSet
./planner/core/memtable_predicate_extractor.go:572:     Instances set.StringSet
./planner/core/memtable_predicate_extractor.go:585:     LogLevels set.StringSet
./planner/core/memtable_predicate_extractor.go:668:     LabelConditions map[string]set.StringSet
./planner/core/memtable_predicate_extractor.go:775:     MetricsNames set.StringSet
./planner/core/memtable_predicate_extractor.go:805:     Rules set.StringSet
./planner/core/memtable_predicate_extractor.go:808:     Items set.StringSet
./planner/core/memtable_predicate_extractor.go:838:     Rules       set.StringSet
./planner/core/memtable_predicate_extractor.go:839:     MetricNames set.StringSet
./planner/core/memtable_predicate_extractor.go:872:     Types       set.StringSet
./planner/core/util.go:242:// extractStringFromStringSet helps extract string info from set.StringSet
./executor/aggregate.go:86:     groupSet            set.StringSet
./executor/aggregate.go:148:    groupSet         set.StringSet
./executor/aggfuncs/func_count_distinct.go:122: valSet set.StringSet
./executor/aggfuncs/func_count_distinct.go:218: valSet set.StringSet
./executor/aggfuncs/func_count_distinct.go:268: valSet set.StringSet
./executor/aggfuncs/func_group_concat.go:155:   valSet            set.StringSet
./executor/aggfuncs/func_group_concat.go:427:   valSet            set.StringSet
./executor/aggfuncs/func_avg.go:193:    valSet set.StringSet
./executor/aggfuncs/func_sum.go:43:     valSet set.StringSet
./executor/inspection_result.go:57:             set       set.StringSet
./expression/builtin_other.go:243:      hashSet set.StringSet
./expression/builtin_other.go:385:      hashSet set.StringSet
```

There are lots of places using `StringSet`, would they all have the same potential issue?

Comment (SunRunAway) [reply executor/aggfuncs/func_count_distinct.go:258]: > how about modifying the `stringmap`?

I'd rather not. The users of `stringmap` should have control of how they handle the memory.

Comment (SunRunAway) [reply executor/aggfuncs/func_count_distinct.go:258]: > There are lots of places using `StringSet`, would they all have the same potential issue?

This PR/issue aims to resolve the problem in package `executor/aggfuncs`, I'll take another look and create issues if there'are further problems and reply here.

**Comment by SunRunAway** [comment]
> Would it impact the performance of these functions?

It has a little impact which cant be seen in benchmark

File: executor/aggfuncs/func_count_distinct.go (hunk 1 above)
Comment (SunRunAway) [reply executor/aggfuncs/func_count_distinct.go:258]: Others have no problem with my glance. And some comments are added.

**Pull request merged**